File: pr1-cs/Server/Servers/ConcurrentServer.cs
using System.Net.Sockets;
using System.Threading;

namespace Server.Servers
{
    public abstract class ConcurrentServer : Server
    {
        protected ConcurrentServer(string host, int port) : base(host, port)
        {
            
        }

        protected override void BeginConversation(TcpClient connection)
        {
            Thread clientProxyThread = CreateClientProxyThread(connection);
            clientProxyThread.Start();
        }

        protected abstract Thread CreateClientProxyThread(TcpClient connection);
    }
}

File: pr1-java/server/src/main/java/pr1Java/server/ConcurrentServer.java
package pr1Java.server;

import java.net.Socket;

public abstract class ConcurrentServer extends Server {

    public ConcurrentServer(Integer port) {
        super(port);
    }

    protected void beginConversation(Socket connection) {
        Configuration.logger.traceEntry("entering with {}", connection);

        Thread clientProxyThread = createClientProxyThread(connection);
        clientProxyThread.start();

        Configuration.logger.traceExit();
    }

    protected abstract Thread createClientProxyThread(Socket connection);
}
